### ISSUE TITLE
Better support for read-only vs. edit modes

### ISSUE THREAD

**Issue by Jermolene** [Jermolene/TiddlyWiki5#1213]
Currently, support for read-only mode is limited to a theme that hides some of the editing apparatus. It's a clumsy arrangement:
- Once read-only mode is engaged it's hard to save changes or perform editing tasks
- It makes it impossible to choose a theme other than the hard-coded Snow White, despite the fact that the changes introduced by the read-only theme could work with any theme
- It doesn't support the scenario of an author having full access but published versions being read-only

It is proposed to replace the current implementation with a new system whereby:
- The wiki has distinct modes for "view" and "edit" (stored in `$:/mode`)
- "view" mode would hide editing features via `<$reveal state="$:/mode">`
- "edit" mode would also optionally make tranclusions visible and directly editable
- Modes can be switched automatically according to customisable rules involving:
  - the URL used to load the wiki
  - the username (if any)

In view mode the following features would be hidden or disabled:
- Page control buttons: new tiddler, new journal, import, set password, save changes
- Tiddler toolbar buttons: edit, clone, new here, new journal here, 
- Importing via drag and drop
- Navigating to missing tiddlers

**Comment by felixhayashi** [comment]
I just noticed it is really confusing when a a website user is asked to create a tiddler if it doesn't exist in read only mode.

![selection_491](https://cloud.githubusercontent.com/assets/4307137/6125348/6732a960-b115-11e4-82b3-faa3480b8ec6.png)

**Comment by pmario** [comment]
IMO read only mode TWs need to check the "missing" tab prior to publishing to prevent this behaviour. ... So there is some more work + knowledge for the editor.

**Comment by Jermolene** [comment]
I think the best solution to making it impossible to navigate links to missing tiddlers might be to not render them as links, but as static text.

**Comment by reflectionalist** [comment]
I was just about to propose supporting exporting to an read-only html and found this.  I really like it!

**Comment by lukehsiao** [comment]
This would really bring TiddlyWiki to a new level. Has there been any updates on this issue?

**Comment by Jermolene** [comment]
Hi @lukehsiao no updates as yet, no. Too much to do!

**Comment by lukehsiao** [comment]
Fair enough. For now, one workaround I've been using is to set the theme as ReadOnly, save the TiddyWiki, then afterwords, edit the HTML file to include this tiddler:
```html
<div created="[timestamp]" creator="[author]" modified="[timestamp]" modifier="[author]" title="$:/state/sidebar">
<pre>no</pre>
</div>
```
Then hosting this file on my public server. Although it still allows viewers to get to the functions if they try, people rarely notice the small << for the sidebar and happily navigate around the Wiki as if it was read-only.

**Comment by Jermolene** [comment]
Hi @lukehsiao there is another trick you can use: make a custom "save changes" button that also adjusts the sidebar state:

```
<$button>
<$action-setfield $tiddler="$:/state/sidebar" $value="no"/>
<$action-sendmessage $message="tm-save-wiki"/>
<$action-setfield $tiddler="$:/state/sidebar" $value="yes"/>
publish changes
</$button>
```

**Comment by danielo515** [comment]
Dear @jermolene,

That's a very neat trick! What about adding a tiddler to TW.com called saving to read only mode

**Comment by lukehsiao** [comment]
@Jermolene I'd love this sort of button. However, I'm admittedly not familiar enough with TiddlyWiki yet to know where I should place that code. Can you give me a pointer?

**Comment by danielo515** [comment]
Just place it in any tiddler. Then open that tiddler when you want to press that button. You can also tag that tiddler appropriately if you want to place it in the sidebar for example

**Comment by lukehsiao** [comment]
@danielo515 So to accomplish the same thing I was doing above, I believe my button should be:
```
<$button>
<$action-setfield $tiddler="$:/state/sidebar" $value="no"/>
<$action-sendmessage $message="tm-save-wiki"/>
publish changes
</$button>
```
However, when I click this button, I see the sidebar disappear and it saves the Wiki, but when I open the wiki again, the sidebar is visible. It doesn't seem to properly make it not visible be default.

**Comment by danielo515** [comment]
Yes, that is correct. How is the TW saved? By downloading it as a new copy or overriding the current tiddler one? Please note that maybe you unhide the sidebar and save again.

**Comment by gernert** [comment]
A long time ago (12/30/2013) a post in the group [1] let me think about making TW5 readonly which resulted in my guide 'TW5 read-only' at http://tw5readonly.tiddlyspot.com/. As of 7/16/2014 there is a readonly plugin.

[1] https://groups.google.com/forum/#!msg/tiddlywiki/hF30Y9WD9lE/YbZHrUK_c2kJ

**Comment by linonetwo** [comment]
I made a simple server example that allows editing only when login via Gitlab

https://github.com/linonetwo/try-passportjs-gitlab-tiddlywiki

I think "edit" means "allow saving to the fs", and "read-only" means "you can edit on-page but not affecting the fs", is enouth.

**Comment by Jermolene** [comment]
Hi @linonetwo that's great, very handy for some common usecases

**Comment by linonetwo** [comment]
https://github.com/linonetwo/tiddlywiki-nodejs-readonly-dockerfile#why-the-default-is-read-only

Readonly by using random password. This is a docker image for deploying readonly lazy-loading blog page.
